noir-projects/aztec-nr/aztec/src/utils/bytes.nr

```diff
@@ -0,0 +1,120 @@
+// Converts the input bytes into an array of fields. A Field is ~254 bits meaning that each field can store 31 bytes.
+// This implies that M = ceil(N / 31).
+pub fn bytes_to_fields<let N: u32, let M: u32>(input: [u8; N]) -> [Field; M] {
+    let mut dst = [0; M];
+
+    for field_index in 0..M {
+        let mut field_value = 0;
+
+        for i in 0..31 {
+            let byte_index = field_index * 31 + i;
+            if byte_index < N {
+                // Shift the existing value left by 8 bits and add the new byte
+                field_value = field_value * 256 + input[byte_index] as Field;
+            }
+        }
+
+        dst[field_index] = field_value;
+    }
+
+    dst
+}
+
+// Converts an input array of fields into bytes. Each field of input has to contain only 31 bytes.
+pub fn fields_to_bytes<let N: u32, let M: u32>(input: [Field; M]) -> [u8; N] {
+    let mut dst = [0; N];
+
+    for field_index in 0..M {
+        // We convert the field to 31 bytes. Normally a field is 254 bits, but we only used 248 bits
+        // in `bytes_to_fields` so we can safely convert it to 31 bytes.
+        let src: [u8; 31] = input[field_index].to_be_bytes();
+
+        // Since some of the bytes might not be occupied (if the source value requiring less than 31 bytes),
+        // we have to compute the start index from which to copy.
+        let remaining_bytes = N - field_index * 31;
+        let src_start_index = if remaining_bytes < 31 {
+            // If the remaining bytes are less than 31, we only copy the remaining bytes
+            31 - remaining_bytes
+        } else {
+            0
+        };
+
+        for i in 0..31 {
+            let byte_index = field_index * 31 + i;
+            if byte_index < N {
+                dst[byte_index] = src[src_start_index + i];
+            }
+        }
+    }
+
+    dst
+}
+
+mod test {
+    use crate::utils::bytes::{bytes_to_fields, fields_to_bytes};
+
+    #[test]
+    fn test_bytes_to_1_field() {
+        let input = [
+            1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20, 21, 22, 23, 24, 25, 26, 27, 28, 29, 30, 31
+        ];
+        let output = bytes_to_fields::<31, 1>(input);
+
+        assert_eq(output[0], 0x0102030405060708090a0b0c0d0e0f101112131415161718191a1b1c1d1e1f);
+    }
+
+    #[test]
+    fn test_1_field_to_bytes() {
```

Comment:
We also need tests for fields being partially used, e.g. `input = [1, 2, 3]`

Reply:
Addressed in https://github.com/AztecProtocol/aztec-packages/pull/8590/commits/7c283d887561f9c791f908c835141d884d25ff8a